Cannot assign to field of Callable type

Consider this example:

``` python
from typing import Callable

class A:
    f = None  # type: Callable[[int], int]

    def __init__(self, g: Callable[[int], int]) -> None:
        self.f = g
```

mypy reports:

```
trial.py: In member "__init__" of class "A":
trial.py, line 7: Cannot assign to a method
trial.py, line 7: Invalid method type
trial.py, line 7: Incompatible types in assignment (expression has type Callable[[int], int], variable has type Callable[[], int])
```

while I'd expect this to work.

Note that there is a second error in the reporting of the type of the variable.

Also, this passes type checking (correctly):

``` python
f = None  # type: Callable[[int], int]

def g(x: int) -> int: return 0

f = g
```

Is there any update? The issue is still present. For some real world example, it is used in the [ctypes](https://docs.python.org/3/library/ctypes.html#function-prototypes) module (in the examples).

```python
from typing import Any, Callable

class A():
    a = lambda: None  # type: Callable[..., None]

def f() -> None: return None

A().a = f
```
gives
```
/tmp/assign_callable.py:8: error: Cannot assign to a method
/tmp/assign_callable.py:8: error: Incompatible types in assignment (expression has type Callable[[], None], variable has type Callable[..., None])
```

Having `Callable[..., None]` or `Callable[[], None]` do not resolve it, having `Callable[..., Any]` remove the incompatible types line.

Hello everybody from July'18

`apps/main/admin.py:120: error: "Callable[[Any, Any], Any]" has no attribute "short_description"`

Django dev love mypy :D

Reading back this issue I'm a little bit confused. The original description (from [2015](https://github.com/python/mypy/issues/708#issue-85773757)!) is about assigning to an instance variable that is backed by a class variable with a Callable type: (1)
```py
class A:
    f = None  # type: Callable[[int], int]

    # The original example had __init__ but it doesn't matter what method it is
    def foo(self, g: Callable[[int], int]) -> None:
        self.f = g  # Three errors here (line 12)
```
The three errors are:
```
_.py:12: error: Cannot assign to a method
_.py:12: error: Invalid self argument "A" to attribute function "f" with type "Callable[[int], int]"
_.py:12: error: Incompatible types in assignment (expression has type "Callable[[int], int]", variable has type "Callable[[], int]")
```
There's a [variation](https://github.com/python/mypy/issues/708#issuecomment-236206421) where the assignment target is `A().f` which elicits essentially the same errors. (2)

But then someone else added a ["me too" comment](https://github.com/python/mypy/issues/708#issuecomment-282351264) that's actually about something completely different (assigning to a function attribute). And that got 8 thumbs up. Those people should really open their own issue. (3)

So excuse me for asking clarification -- which of the above three do you (@glyph) need help finding a workaround for? I was going to suggest putting `# type: ignore` on the assignment that gets the error, but then I realized that the problem is just as much *using* the `f` attribute! This is also an error:
```py
A().f(42)
```
The errors:
```
_.py:16: error: Invalid self argument "A" to attribute function "f" with type "Callable[[int], int]"
_.py:16: error: Too many arguments
```
This was also shown in [Jukka's comment](https://github.com/python/mypy/issues/708#issuecomment-109783888) but perhaps not emphasized enough (and he confused matters by considering a callable that was an instance of a class with a `__call__` method).

Now what to do about it? The best I can come up with is this:
```py
class A:
    if TYPE_CHECKING:
        def f(self, arg: int) -> int:
            ...
    else:
        f = None

    def foo(self, g: Callable[[int], int]) -> None:
        self.f = g  # type: ignore
```
This makes calling `A().f(42)` pass without errors (and return the right type).

I'm not sure if this comparison will be helpful, but I just checked pyright to see what it does in this situation (since I've noticed that it handles class-level vs instance level type hinting for instance properties in a different way).  It turns out that pyright is 100% ok with class-level hints of callable instance-level properties.

To demonstrate that, here is the same code as above, tuned up very slightly to avoid the None assignment at class level:

```
from typing import Callable

class A:
    f: Callable[[int], int]  # <-- instance property type included at class-level (w/o None)
    def __init__(self, g: Callable[[int], int]) -> None:
        self.f = g

# Test the  usage for good measure...
def plus1(x: int) -> int:
    return x + 1
a = A(plus1)
print(a.f(1.1))  # <-- pyright's only complaint is this float (which is what we want)
print(a.f(1))
```

And here are the outputs for `mypy` and `pyright`:

```
$ python --version
Python 3.8.2
$ mypy --version
mypy 0.770
$ mypy --strict ./hmm.py 
hmm.py:6: error: Cannot assign to a method
hmm.py:6: error: Invalid self argument "A" to attribute function "f" with type "Callable[[int], int]"
hmm.py:6: error: Incompatible types in assignment (expression has type "Callable[[int], int]", variable has type "Callable[[], int]")
hmm.py:12: error: Invalid self argument "A" to attribute function "f" with type "Callable[[int], int]"
hmm.py:12: error: Too many arguments
hmm.py:13: error: Invalid self argument "A" to attribute function "f" with type "Callable[[int], int]"
hmm.py:13: error: Too many arguments
Found 7 errors in 1 file (checked 1 source file)
$ 
$ 
$ pyright --version
pyright 1.1.29
$ pyright ./hmm.py 
typingsPath /home/russ/hmm/typings is not a valid directory.
Searching for source files
Found 1 source file
/home/russ/hmm/hmm.py
  12:11 - error: Argument of type 'Literal[1.1]' cannot be assigned to parameter 'p0' of type 'int'
  'float' is incompatible with 'int'
1 error, 0 warnings 
Completed in 1.474sec
```

mypy has 7 complaints about the code (including the one that originated this issue), while pyright only complains about the legitimately bad float usage.

I still hesitantly prefer the workaround (?) in my previous post where you skip the hint at class level entirely, since A) it isn't a class level property, and B) you don't need to type out the exact same hint for the `__init__` arg.  However, I say "hesitantly", because (per previous) I'm not sure what the impact is of violating PEP526's recommendation to define instance property types at class level.

--- a/mypy/checkmember.py
+++ b/mypy/checkmember.py
@@ -530,6 +530,15 @@ def instance_alias_type(alias: TypeAlias,
     tp = type_object_type(target.type, builtin_type)
     return expand_type_by_instance(tp, target)
 
+def is_instance_var(var: Var, info: TypeInfo) -> bool:
+    """Return if var is an instance variable according to PEP 526."""
+    return (
+        # check the type_info node is the var (not a decorated function, etc.)
+        var.name in info.names and info.names[var.name].node is var
+        and not var.is_classvar
+        # variables without annotations are treated as classvar
+        and not var.is_inferred
+    )
 
 def analyze_var(name: str,
                 var: Var,
@@ -559,7 +568,12 @@ def analyze_var(name: str,
         t = get_proper_type(expand_type_by_instance(typ, itype))
         result = t  # type: Type
         typ = get_proper_type(typ)
-        if var.is_initialized_in_class and isinstance(typ, FunctionLike) and not typ.is_type_obj():
+        if (
+            var.is_initialized_in_class
+            and not is_instance_var(var, info)
+            and isinstance(typ, FunctionLike)
+            and not typ.is_type_obj()
+        ):
             if mx.is_lvalue:
                 if var.is_property:
                     if not var.is_settable_property:

--- a/mypy/checkmember.py
+++ b/mypy/checkmember.py
@@ -530,6 +530,7 @@ def instance_alias_type(alias: TypeAlias,
     tp = type_object_type(target.type, builtin_type)
     return expand_type_by_instance(tp, target)
 
+
 def is_instance_var(var: Var, info: TypeInfo) -> bool:
     """Return if var is an instance variable according to PEP 526."""
     return (
@@ -539,6 +540,7 @@ def is_instance_var(var: Var, info: TypeInfo) -> bool:
         # variables without annotations are treated as classvar
         and not var.is_inferred
     )
+
 
 def analyze_var(name: str,
                 var: Var,

--- a/test-data/unit/check-dataclasses.test
+++ b/test-data/unit/check-dataclasses.test
@@ -1109,71 +1109,37 @@ reveal_type(B)  # N: Revealed type is "def (foo: builtins.int) -> __main__.B"
 
 [builtins fixtures/property.pyi]
 
-[case testDataclassCallableProperty]
+[case testDataclassCallableFieldAccess]
 # flags: --python-version 3.7
 from dataclasses import dataclass
 from typing import Callable
 
 @dataclass
 class A:
-    foo: Callable[[int], int]
+    x: Callable[[int], int]
+    y: Callable[[int], int] = lambda i: i
 
-def my_foo(x: int) -> int:
-    return x
+a = A(lambda i:i)
+x: int = a.x(0)
+y: str = a.y(0) # E: Incompatible types in assignment (expression has type "int", variable has type "str")
+reveal_type(a.x)  # N: Revealed type is "def (builtins.int) -> builtins.int"
+reveal_type(a.y)  # N: Revealed type is "def (builtins.int) -> builtins.int"
+reveal_type(A.y)  # N: Revealed type is "def (builtins.int) -> builtins.int"
 
-a = A(foo=my_foo)
-a.foo(1)
-reveal_type(a.foo)  # N: Revealed type is "def (builtins.int) -> builtins.int"
-reveal_type(A.foo)  # N: Revealed type is "def (builtins.int) -> builtins.int"
-[typing fixtures/typing-medium.pyi]
-[case testDataclassCallableAssignment]
+[case testDataclassCallableFieldAssignment]
 # flags: --python-version 3.7
 from dataclasses import dataclass
 from typing import Callable
 
 @dataclass
 class A:
-    foo: Callable[[int], int]
+    x: Callable[[int], int]
 
-def my_foo(x: int) -> int:
-    return x
+def x(i: int) -> int:
+    return i
+def x2(s: str) -> str:
+    return s
 
-a = A(foo=my_foo)
-
-def another_foo(x: int) -> int:
-    return x + 1
-
-a.foo = another_foo
-[case testDataclassCallablePropertyWrongType]
-# flags: --python-version 3.7
-from dataclasses import dataclass
-from typing import Callable
-
-@dataclass
-class A:
-    foo: Callable[[int], int]
-
-def my_foo(x: int) -> str:
-    return "foo"
-
-a = A(foo=my_foo)  # E: Argument "foo" to "A" has incompatible type "Callable[[int], str]"; expected "Callable[[int], int]"
-[typing fixtures/typing-medium.pyi]
-[case testDataclassCallablePropertyWrongTypeAssignment]
-# flags: --python-version 3.7
-from dataclasses import dataclass
-from typing import Callable
-
-@dataclass
-class A:
-    foo: Callable[[int], int]
-
-def my_foo(x: int) -> int:
-    return x
-
-a = A(foo=my_foo)
-
-def another_foo(x: int) -> str:
-    return "foo"
-
-a.foo = another_foo  # E: Incompatible types in assignment (expression has type "Callable[[int], str]", variable has type "Callable[[int], int]")
-[typing fixtures/typing-medium.pyi]
+a = A(lambda i:i)
+a.x = x
+a.x = x2 # E: Incompatible types in assignment (expression has type "Callable[[str], str]", variable has type "Callable[[int], int]")

--- a/test-data/unit/check-functions.test
+++ b/test-data/unit/check-functions.test
@@ -621,6 +621,18 @@ a = None # type: A
 a.f()
 a.f(a) # E: Too many arguments
 
+[case testMethodWithInferredMethodAsDataAttribute]
+from typing import Any
+def m(self: "A") -> int: ...
+
+class A:
+    n = m
+
+a = A()
+reveal_type(a.n())  # N: Revealed type is "builtins.int"
+reveal_type(A.n(a))  # N: Revealed type is "builtins.int"
+A.n()  # E: Too few arguments
+
 [case testOverloadedMethodAsDataAttribute]
 from foo import *
 [file foo.pyi]

--- a/test-data/unit/check-functions.test
+++ b/test-data/unit/check-functions.test
@@ -588,12 +588,12 @@ A().f('') # E: Argument 1 to "f" of "A" has incompatible type "str"; expected "i
 
 
 [case testMethodAsDataAttribute]
-from typing import Any, Callable
+from typing import Any, Callable, ClassVar
 class B: pass
 x = None # type: Any
 class A:
-    f = x # type: Callable[[A], None]
-    g = x # type: Callable[[A, B], None]
+    f = x # type: ClassVar[Callable[[A], None]]
+    g = x # type: ClassVar[Callable[[A, B], None]]
 a = None # type: A
 a.f()
 a.g(B())
@@ -601,22 +601,22 @@ a.f(a) # E: Too many arguments
 a.g()  # E: Too few arguments
 
 [case testMethodWithInvalidMethodAsDataAttribute]
-from typing import Any, Callable
+from typing import Any, Callable, ClassVar
 class B: pass
 x = None # type: Any
 class A:
-    f = x # type: Callable[[], None]
-    g = x # type: Callable[[B], None]
+    f = x # type: ClassVar[Callable[[], None]]
+    g = x # type: ClassVar[Callable[[B], None]]
 a = None # type: A
 a.f() # E: Attribute function "f" with type "Callable[[], None]" does not accept self argument
 a.g() # E: Invalid self argument "A" to attribute function "g" with type "Callable[[B], None]"
 
 [case testMethodWithDynamicallyTypedMethodAsDataAttribute]
-from typing import Any, Callable
+from typing import Any, Callable, ClassVar
 class B: pass
 x = None # type: Any
 class A:
-    f = x # type: Callable[[Any], Any]
+    f = x # type: ClassVar[Callable[[Any], Any]]
 a = None # type: A
 a.f()
 a.f(a) # E: Too many arguments
@@ -662,35 +662,35 @@ a.g(B())
 a.g(a)   # E: Argument 1 has incompatible type "A[B]"; expected "B"
 
 [case testInvalidMethodAsDataAttributeInGenericClass]
-from typing import Any, TypeVar, Generic, Callable
+from typing import Any, TypeVar, Generic, Callable, ClassVar
 t = TypeVar('t')
 class B: pass
 class C: pass
 x = None # type: Any
 class A(Generic[t]):
-    f = x # type: Callable[[A[B]], None]
+    f = x # type: ClassVar[Callable[[A[B]], None]]
 ab = None # type: A[B]
 ac = None # type: A[C]
 ab.f()
 ac.f()   # E: Invalid self argument "A[C]" to attribute function "f" with type "Callable[[A[B]], None]"
 
 [case testPartiallyTypedSelfInMethodDataAttribute]
-from typing import Any, TypeVar, Generic, Callable
+from typing import Any, TypeVar, Generic, Callable, ClassVar
 t = TypeVar('t')
 class B: pass
 class C: pass
 x = None # type: Any
 class A(Generic[t]):
-    f = x # type: Callable[[A], None]
+    f = x # type: ClassVar[Callable[[A], None]]
 ab = None # type: A[B]
 ac = None # type: A[C]
 ab.f()
 ac.f()
 
 [case testCallableDataAttribute]
-from typing import Callable
+from typing import Callable, ClassVar
 class A:
-    g = None # type: Callable[[A], None]
+    g = None # type: ClassVar[Callable[[A], None]]
     def __init__(self, f: Callable[[], None]) -> None:
         self.f = f
 a = A(None)

--- a/test-data/unit/check-functools.test
+++ b/test-data/unit/check-functools.test
@@ -25,12 +25,12 @@ Ord() >= 1  # E: Unsupported operand types for >= ("Ord" and "int")
 
 [case testTotalOrderingLambda]
 from functools import total_ordering
-from typing import Any, Callable
+from typing import Any, Callable, ClassVar
 
 @total_ordering
 class Ord:
-    __eq__: Callable[[Any, object], bool] = lambda self, other: False
-    __lt__: Callable[[Any, "Ord"], bool] = lambda self, other: False
+    __eq__: ClassVar[Callable[[Any, object], bool]] = lambda self, other: False
+    __lt__: ClassVar[Callable[[Any, "Ord"], bool]] = lambda self, other: False
 
 reveal_type(Ord() < Ord())  # N: Revealed type is "builtins.bool"
 reveal_type(Ord() <= Ord())  # N: Revealed type is "builtins.bool"

--- a/test-data/unit/check-selftype.test
+++ b/test-data/unit/check-selftype.test
@@ -366,7 +366,7 @@ reveal_type(x.f)  # N: Revealed type is "builtins.int"
 [builtins fixtures/property.pyi]
 
 [case testSelfTypeProperSupertypeAttribute]
-from typing import Callable, TypeVar
+from typing import Callable, TypeVar, ClassVar
 class K: pass
 T = TypeVar('T', bound=K)
 class A(K):
@@ -374,8 +374,8 @@ class A(K):
     def g(self: K) -> int: return 0
     @property
     def gt(self: T) -> T: return self
-    f: Callable[[object], int]
-    ft: Callable[[T], T]
+    f: ClassVar[Callable[[object], int]]
+    ft: ClassVar[Callable[[T], T]]
 
 class B(A):
     pass
@@ -392,15 +392,15 @@ reveal_type(B().ft())  # N: Revealed type is "__main__.B*"
 [builtins fixtures/property.pyi]
 
 [case testSelfTypeProperSupertypeAttributeTuple]
-from typing import Callable, TypeVar, Tuple
+from typing import Callable, TypeVar, Tuple, ClassVar
 T = TypeVar('T')
 class A(Tuple[int, int]):
     @property
     def g(self: object) -> int: return 0
     @property
     def gt(self: T) -> T: return self
-    f: Callable[[object], int]
-    ft: Callable[[T], T]
+    f: ClassVar[Callable[[object], int]]
+    ft: ClassVar[Callable[[T], T]]
 
 class B(A):
     pass
@@ -450,7 +450,7 @@ reveal_type(X1.ft())  # N: Revealed type is "Type[__main__.X]"
 [builtins fixtures/property.pyi]
 
 [case testSelfTypeProperSupertypeAttributeGeneric]
-from typing import Callable, TypeVar, Generic
+from typing import Callable, TypeVar, Generic, ClassVar
 Q = TypeVar('Q', covariant=True)
 class K(Generic[Q]):
     q: Q
@@ -460,8 +460,8 @@ class A(K[Q]):
     def g(self: K[object]) -> int: return 0
     @property
     def gt(self: K[T]) -> T: return self.q
-    f: Callable[[object], int]
-    ft: Callable[[T], T]
+    f: ClassVar[Callable[[object], int]]
+    ft: ClassVar[Callable[[T], T]]
 
 class B(A[Q]):
     pass

